### PR TITLE
[AIP-122] Keys must be snake case.

### PR DESCRIPTION
**Pull request by lukesneeringer**
Also, make the tests better.

### CODE REVIEW

**Comment by lukesneeringer** [comment]
Marking `do not merge`; this still has one bug.

File: rules/aip0122/http_url_pattern.go [hunk 1]
```diff
@@ -23,19 +23,56 @@ import (
...
-					Message:    "HTTP URL should use camel case, but not snake case.",
+			parsedURI := parseURI(httpRule.URI)
+			if strings.Contains(parsedURI.pattern, "_") {
+				problems = append(problems, lint.Problem{
+					Message:    "HTTP URI patterns should use camel case, not snake case.",
 					Descriptor: m,
-				}}
+				})
+			}
+			for _, v := range parsedURI.vars {
```
Comment (mingzhi) [inline rules/aip0122/http_url_pattern.go]: Shouldn't we create a separate rule for checking variable names?

File: rules/aip0122/http_url_pattern.go [hunk 2]
```diff
@@ -23,19 +23,56 @@ import (
...
-		return nil
+		return
 	},
 }
+
+// extractURI separates variable names from the URI pattern.
+//
+// It does not maintain the relationship between variable names and what
+// part of the pattern they apply to (we do not need that to lint here).
+func parseURI(uri string) (parsed struct {
```
Comment (mingzhi) [inline rules/aip0122/http_url_pattern.go]: This function contains complicated logics -- should we create a unit test for it?

Comment (lukesneeringer) [reply rules/aip0122/http_url_pattern.go]: In general, you should not test private (unexported) functions. I ensured that the existing test hits every statement.

File: rules/aip0122/http_url_pattern.go (hunk 1 above)
Comment (lukesneeringer) [reply rules/aip0122/http_url_pattern.go]: It is fundamentally the same string, just different pieces of it. I could be convinced either way.

**Comment by lukesneeringer** [comment]
Fixed my bug.

File: rules/aip0122/http_url_pattern.go (hunk 1 above)
Comment (lukesneeringer) [reply rules/aip0122/http_url_pattern.go]: I _think_ I very slightly prefer a single rule. These really are part and parcel the same thing.

Comment (mingzhi) [reply rules/aip0122/http_url_pattern.go]: Okay, thanks!

File: rules/aip0122/http_url_pattern.go (hunk 2 above)
Comment (mingzhi) [reply rules/aip0122/http_url_pattern.go]: Okay.

**Pull request merged**